text/3389-manifest-lint.md

```diff
@@ -0,0 +1,250 @@
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controling precedence between lints and lint groups?  We are
+using a TOML table with the keys as lint names which does not allow controlling
+ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
+precedence between levels that the user can't control.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Configurable lints
+
+We can extend basic lint syntax:
+```toml
+[lints]
+cyclomatic_complexity = "allow"
+```
+to support configuration, whether for cargo or the lint tool:
+```toml
+[lints]
+cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }
+```
+Where `rust-version` is used by cargo to determine whether to pass along this
+lint and `threshold` is used by the tool.  We'd need to define how to
+distinguish between reserved and unreserved field names.
+
+## Extending the syntax to `.cargo/config.toml`
```

Comment:
I actually think this is probably necessary for this RFC to make sense. I feel this makes more sense in the local (potentially-but-not-necessarily checked-in) config as opposed to the crate config.

Reply:
Though with workspace inheritance perhaps it's not that necessary.

Reply:
Could you clarify why you feel it makes more sense to put this in `.cargo/config.toml` over `Cargo.toml`?

In starting this with `Cargo.toml`, I'm looking to address the existing problems with doing this via rustflags in `.cargo/config.toml` (project configuration being directory dependent).  https://github.com/EmbarkStudios/rust-ecosystem/issues/59 is a good example of people using both source code and `.cargo/config.toml` for lints that are project-specific and would benefit from this.   This was also spurred by the cargo team itself wanting to use this feature.

Reply:
I think it makes sense to allow it in _both_.

The reason being that there are multiple lint workflows: sometimes you lint in CI and want everyone to follow the same lint set. However, some people may wish to opt in to additional lints locally for their own code, and being able to set that is useful.

But I'm fine with starting out in Cargo.toml at least. I would _prefer_ to try and solve it for both.

Reply:
> The reason being that there are multiple lint workflows: sometimes you lint in CI and want everyone to follow the same lint set. However, some people may wish to opt in to additional lints locally for their own code, and being able to set that is useful.

The part I'm missing is why users can't just use rustflags for this in the short term?  In the long term, if we add configurable lints, there isn't an alternative and I can see it becoming a higher priority then.  This is the main reason I put it as a future possibility, rather than including it.

Reply:
I don't think we should add this to config.toml.

People can still do this via rustflags if they really want to, but I don't think we should encourage a workflow in which some lints are enabled in some build contexts but not in others.

Reply:
> @joshtriplett I don't think we should encourage a workflow in which some lints are enabled in some build contexts but not in others.

So you're saying that I shouldn't be able to easily enable custom lints on my machine if the projects don't already use them?

There are many Clippy lints which have high false positive rates, too restrictive or even self-contradictory. `pedantic` and `restriction` categories are full of those. They are infeasible to gate CI on, which means they shouldn't be enabled in the project's source or `Cargo.toml`. But there is no reason to not allow easily running those locally. Neither should we make it hard to locally disable any lints, including all warnings. The quality control is enforced by CI anyway, and if I'm digging in the Git history, lint warnings are just syntactic noise which can't be acted on anyway.

The rustflags solution specifically is bad, because setting it blows the build cache. Again, if I'm digging Git history, or checking out someone else's branch, or just taking a look at some random project on Github, I'd usually want custom lint settings (sometimes more lax than the defaults, sometimes more strict). Setting rustflags would mean that all my cached dependencies would need to be recompiled, for no good reason. It breaks incremental compilation, increases compile times, increases power usage and SSD wear, and all for no reason other than Cargo maintainers' personal taste. I could easily set those for rustc if I used a different build system, the compiler already supports it, it's just that Cargo doesn't want to make it easy to use.

Reply:
> The rustflags solution specifically is bad, because setting it blows the build cache. Again, if I'm digging Git history, or checking out someone else's branch, or just taking a look at some random project on Github, I'd usually want custom lint settings (sometimes more lax than the defaults, sometimes more strict). Setting rustflags would mean that all my cached dependencies would need to be recompiled, for no good reason. 

This wasn't too clear to me.  We will still be fingerprinting `[lints]` and causing rebuilds, so having support for `[lints]` in `.cargo/config.toml` will be no different for rebuilds than rustflags.  Well, we might specialize the fingerprinting for the tool (e.g. don't fingerprint clippy lints for rustc builds).  Is that the part that you feel would be an improvement?

Reply:
When you are cross-compiling, there is no way to set rustflags via `.config/cargo.toml` that applies *to everything* (built for host or built for target), so even just for having something that works across proc-macros and normal crates within a workspace, this would be great.

Reply:
> We will still be fingerprinting [lints] and causing rebuilds

Why? Lints don't affect the compiled binaries. It seems like pure overhead to me.

----- 

While I'd want to be able to set lints globally (for every project on my computer), there is a design issue of config priority which is unclear to me. Currently afaik the only option duplicated in both `config.toml` and `Cargo.toml` is [`[profile]`](https://doc.rust-lang.org/cargo/reference/config.html#profile), and in its case the `config.toml` has priority. In particular, a `[profile]` section in `$HOME/.cargo/config.toml` would override the `[profile]` in any other Cargo project.

That may be reasonable for `[profile]`, but would likely be incorrect for lints. Most of the time I'd want to use the lints from the specific project I work on, with the `config.toml` defaults (particularly defined in non-project directories) used only occasionally. So the correct priority looks like `$WORKSPACE/.cargo/config.toml` > `$WORKSPACE/subcrate/Cargo.toml` > `$WORKSPACE/Cargo.toml` > `$UPPER_DIR/.cargo/config.toml`. This looks a bit convoluted and inconsistent with the behaviour of other `config.toml` keys.

Maybe there should be a cargo command-line flag which would control whether `Cargo.toml` or `config.toml` lint settings should be used?

Actually, if `cargo build/run/test` would have flags to set lint levels per invocation, that would be sufficient. That would also avoid the question of config priorities.

Reply:
> Why? Lints don't affect the compiled binaries. It seems like pure overhead to me.

If a user adds additional lints and run `cargo check`, they would expect to see the result of those lints.  This would only be possible by rebuilding which will require finger printing.

> While I'd want to be able to set lints globally (for every project on my computer), there is a design issue of config priority which is unclear to me. 

For now, I am leaving `[lints]` for `.cargo/config.toml` as a future possibility rather than removing it or making it required to move this proposal along.     Keeping it as a future possibility has no cost.  Adding it into the proposal has a larger cost, in design, decision making, and implementation.  It would require a very strong motivation to block incremental improvement (`[lints]` in `Cargo.toml`) on getting a full solution (e.g. including `[lints]` in `.cargo/config.toml`).

With that context, maybe we can temper how much design work we put into the feature or how much we debate whether we should or shouldn't do it in the future.

Reply:
Sure, I wouldn't want future extensions to block this long-waited feature either. 

When you say "rebuilding", do you mean rebuilding the workspace crates or all dependencies? I am most concerned about the latter, even though for large projects there could be no difference. I am not familiar with the inner details of `cargo check`.